File: src/main/kotlin/com/jakewharton/gitout/GitHub.kt
package com.jakewharton.gitout

import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.api.Optional
import com.apollographql.apollo.network.http.DefaultHttpEngine
import dev.drewhamilton.poko.Poko
import okhttp3.OkHttpClient

internal class GitHub(
	private val user: String,
	private val token: String,
	private val client: OkHttpClient,
	private val logger: Logger,
) {
	@Poko
	class UserRepositories(
		val owned: Set<String>,
		val starred: Set<String>,
		val watching: Set<String>,
		val gists: Set<String>,
		val metadata: Map<String, RepositoryMetadata>,
	)

	suspend fun loadRepositories(): UserRepositories {
		val owned = mutableSetOf<String>()
		val starred = mutableSetOf<String>()
		val watching = mutableSetOf<String>()
		val gists = mutableSetOf<String>()
		val metadata = mutableMapOf<String, RepositoryMetadata>()

		ApolloClient.Builder()
			.serverUrl("https://api.github.com/graphql")
			.httpEngine(DefaultHttpEngine(client))
			.addHttpHeader("Authorization", "Bearer $token")
			.addHttpHeader("User-Agent", "gitout/$version")
			.build()
			.use { client ->
				var ownedAfter: Optional<String?> = Optional.Absent
				var starredAfter: Optional<String?> = Optional.Absent
				var watchingAfter: Optional<String?> = Optional.Absent
				var gistsAfter: Optional<String?> = Optional.Absent

				while (true) {
					val query = UserReposQuery(
						login = user,
						ownerAfter = ownedAfter,
						starredAfter = starredAfter,
						watchingAfter = watchingAfter,
						gistsAfter = gistsAfter,
					)
					logger.debug { "Requesting page $query" }

					val response = client.query(query).execute().dataAssertNoErrors
					logger.trace { response.toString() }

					val user = response.user
			?: throw IllegalStateException("GitHub user not found: $user. Check username and token permissions.")

					val ownedEdges = user.ownedRepositories.ownedEdges.orEmpty()
					val starredEdges = user.starredRepositories.starredEdges.orEmpty()
					val watchingEdges = user.watchingRepositories.watchingEdges.orEmpty()
					val gistEdges = user.gistRepositories.gistEdges.orEmpty()
					if (ownedEdges.isEmpty() && starredEdges.isEmpty() && watchingEdges.isEmpty() && gistEdges.isEmpty()) {
						break // All cursors exhausted!
					}

					for (ownedEdge in ownedEdges) {
						val node = ownedEdge?.node
						if (node != null) {
							owned += node.nameWithOwner
							metadata[node.nameWithOwner] = RepositoryMetadata(
								name = node.nameWithOwner,
								isArchived = node.isArchived,
								isPrivate = node.isPrivate,
								isFork = node.isFork,
								visibility = node.visibility.name,
								description = node.description,
								updatedAt = node.updatedAt.toString(),
								repoType = "owned",
								diskUsageKb = node.diskUsage?.toLong(),
							)
							ownedAfter = Optional.present(ownedEdge.cursor)
						} else {
							logger.warn("Skipping owned repository with null node")
						}
					}
					for (starredEdge in starredEdges) {
						val node = starredEdge?.node
						if (node != null) {
							starred += node.nameWithOwner
							// Only add metadata if not already present (owned takes priority)
							if (node.nameWithOwner !in metadata) {
								metadata[node.nameWithOwner] = RepositoryMetadata(
									name = node.nameWithOwner,
									isArchived = node.isArchived,
									isPrivate = node.isPrivate,
									isFork = node.isFork,
									visibility = node.visibility.name,
									description = node.description,
									updatedAt = node.updatedAt.toString(),
									repoType = "starred",
									diskUsageKb = node.diskUsage?.toLong(),
								)
							}
							starredAfter = Optional.present(starredEdge.cursor)
						} else {
							logger.warn("Skipping starred repository with null node")
						}
					}
					for (watchingEdge in watchingEdges) {
						val node = watchingEdge?.node
						if (node != null) {
							watching += node.nameWithOwner
							// Only add metadata if not already present
							if (node.nameWithOwner !in metadata) {
								metadata[node.nameWithOwner] = RepositoryMetadata(
									name = node.nameWithOwner,
									isArchived = node.isArchived,
									isPrivate = node.isPrivate,
									isFork = node.isFork,
									visibility = node.visibility.name,
									description = node.description,
									updatedAt = node.updatedAt.toString(),
									repoType = "watching",
									diskUsageKb = node.diskUsage?.toLong(),
								)
							}
							watchingAfter = Optional.present(watchingEdge.cursor)
						} else {
							logger.warn("Skipping watching repository with null node")
						}
					}
					for (gistEdge in gistEdges) {
						val node = gistEdge?.node
						if (node != null) {
							gists += node.name
							metadata[node.name] = RepositoryMetadata(
								name = node.name,
								isArchived = false, // Gists don't have archive status
								isPrivate = !node.isPublic,
								isFork = false, // Gists don't have fork status
								visibility = if (node.isPublic) "PUBLIC" else "PRIVATE",
								description = node.description,
								updatedAt = node.updatedAt.toString(),
								repoType = "gist",
							)
							gistsAfter = Optional.present(gistEdge.cursor)
						} else {
							logger.warn("Skipping gist with null node")
						}
					}
				}
			}

		return UserRepositories(
			owned = owned,
			starred = starred,
			watching = watching,
			gists = gists,
			metadata = metadata,
		)
	}
}

File: python/gitout/github.py
"""GitHub repository discovery via the GraphQL API.

Port of ``GitHub.kt``. The data containers are the spec; ``parse_user_repositories``
(the pure paging/dedupe logic over GraphQL ``user`` payloads) is implemented in Phase 1.
The query document is reused verbatim from ``src/main/graphql/GitHub.graphql``.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass(frozen=True)
class RepositoryMetadata:
    name: str
    is_archived: bool
    is_private: bool
    is_fork: bool
    visibility: str
    description: str | None
    updated_at: str
    repo_type: str  # "owned" | "starred" | "watching" | "gist"
    disk_usage_kb: int | None = None
    default_branch: str | None = None
    topics: list[str] = field(default_factory=list)
    language: str | None = None


@dataclass(frozen=True)
class UserRepositories:
    owned: set[str]
    starred: set[str]
    watching: set[str]
    gists: set[str]
    metadata: dict[str, RepositoryMetadata]


def _edges(connection: dict[str, Any] | None, key: str) -> list[dict[str, Any]]:
    if not connection:
        return []
    return connection.get(key) or []


def _repo_metadata(node: dict[str, Any], repo_type: str) -> RepositoryMetadata:
    topic_nodes = (node.get("repositoryTopics") or {}).get("nodes") or []
    topics = [t["topic"]["name"] for t in topic_nodes if t and t.get("topic")]
    default_branch_ref = node.get("defaultBranchRef")
    primary_language = node.get("primaryLanguage")
    return RepositoryMetadata(
        name=node["nameWithOwner"],
        is_archived=node["isArchived"],
        is_private=node["isPrivate"],
        is_fork=node["isFork"],
        visibility=node["visibility"],
        description=node.get("description"),
        updated_at=node["updatedAt"],
        repo_type=repo_type,
        disk_usage_kb=node.get("diskUsage"),
        default_branch=default_branch_ref["name"] if default_branch_ref else None,
        topics=topics,
        language=primary_language["name"] if primary_language else None,
    )


def _gist_metadata(node: dict[str, Any]) -> RepositoryMetadata:
    is_public = node["isPublic"]
    return RepositoryMetadata(
        name=node["name"],
        is_archived=False,  # gists have no archive status
        is_private=not is_public,
        is_fork=False,  # gists have no fork status
        visibility="PUBLIC" if is_public else "PRIVATE",
        description=node.get("description"),
        updated_at=node["updatedAt"],
        repo_type="gist",
    )


def parse_user_repositories(pages: list[dict[str, Any]]) -> UserRepositories:
    """Fold successive GraphQL ``data`` payloads into a :class:`UserRepositories`.

    Each element of ``pages`` is one query response's ``data`` object (containing a
    ``user`` key). Metadata dedupes with owned > starred > watching priority; owned
    and gist metadata overwrite unconditionally. Mirrors ``GitHub.loadRepositories``.
    """
    owned: set[str] = set()
    starred: set[str] = set()
    watching: set[str] = set()
    gists: set[str] = set()
    metadata: dict[str, RepositoryMetadata] = {}

    for page in pages:
        user = page.get("user")
        if user is None:
            continue

        for edge in _edges(user.get("ownedRepositories"), "ownedEdges"):
            node = edge.get("node") if edge else None
            if node:
                owned.add(node["nameWithOwner"])
                metadata[node["nameWithOwner"]] = _repo_metadata(node, "owned")

        for edge in _edges(user.get("starredRepositories"), "starredEdges"):
            node = edge.get("node") if edge else None
            if node:
                starred.add(node["nameWithOwner"])
                if node["nameWithOwner"] not in metadata:
                    metadata[node["nameWithOwner"]] = _repo_metadata(node, "starred")

        for edge in _edges(user.get("watchingRepositories"), "watchingEdges"):
            node = edge.get("node") if edge else None
            if node:
                watching.add(node["nameWithOwner"])
                if node["nameWithOwner"] not in metadata:
                    metadata[node["nameWithOwner"]] = _repo_metadata(node, "watching")

        for edge in _edges(user.get("gistRepositories"), "gistEdges"):
            node = edge.get("node") if edge else None
            if node:
                gists.add(node["name"])
                metadata[node["name"]] = _gist_metadata(node)

    return UserRepositories(
        owned=owned,
        starred=starred,
        watching=watching,
        gists=gists,
        metadata=metadata,
    )
